### ISSUE TITLE
Huge performance difference between keras.models.load_model() and model.load_weights()

### ISSUE THREAD

**Issue by JustASquid** [keras-team/keras#17652]
**System information**.
- Have I written custom code (as opposed to using a stock example script provided in Keras):
Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
Tested in OSX 13.2.1 (M1 Pro) as well as Windows 10 (Nvidia GPU)
- TensorFlow installed from (source or binary):
Binary
- TensorFlow version (use command below):
2.11
- Python version: 3.9
- GPU model and memory: 
OSX: M1 Pro (tensorflow-metal installed; same issue applies when running CPU only)
Windows: 1080 TI

**Describe the problem**.

Consider this code:
```
from keras.applications import InceptionResNetV2

# Takes 68 seconds on M1 Pro
def test_save_inception():
    model = InceptionResNetV2(weights=None)
    model_path = Path("models/inception")
    model.save(model_path)


# Takes 47 seconds on M1 Pro
def test_load_inception():
    model_path = Path("models/test_inception")
    model = tf.keras.models.load_model(model_path, custom_objects={
        "InceptionResNetV2": InceptionResNetV2
    })
    model.summary()


# Takes 9 seconds on M1 Pro
def test_load_inception_2():
    model_path = Path("models/inception")
    model = InceptionResNetV2(weights=None)
    model.load_weights(model_path)
    model.summary()
```

**Describe the current behavior**.

After running `test_save_inception()`, running `test_load_inception()` is much, much slower than `test_load_inception_2()`, even though the resulting models are the same.

**Describe the expected behavior**.

I would expect the implementation of `test_load_inception()` to behave very similarly to `test_load_inception_2()`, because the custom object is provided. However it seems that in reality, providing the custom object doesn't actually make any difference to the loading time or the output model.

**Comment by NiharJani2002** [comment]
@tilakrayal. I am giving it a try for the solution.
@JustASquid. review, my solution.

The issue that the original code is facing is related to the way in which the model is saved and loaded. Specifically, the test_load_inception() function is much slower than the test_load_inception_2() function because it has to reconstruct the entire model architecture when loading the saved model.

When we save a Keras model using the model.save() function, it saves the entire model, including its architecture, weights, and optimizer state (if any). This means that when we load the saved model using tf.keras.models.load_model(), we are loading the entire model including its architecture, weights, and optimizer state. This can be a computationally expensive process, especially for larger models like InceptionResNetV2.

The test_load_inception_2() function takes a different approach to loading the model. Instead of loading the entire model using tf.keras.models.load_model(), it loads only the model weights using model.load_weights(). This approach does not include the optimizer state, but it does not require loading the entire model architecture again. Instead, it constructs the model architecture separately using InceptionResNetV2(weights=None).

The reason why test_load_inception() is slower than test_load_inception_2() is that it has to reconstruct the entire model architecture when loading the saved model. This means that it has to create all of the layers and connections in the model from scratch, which can be computationally expensive.

To improve the performance of test_load_inception(), we can modify it to load only the model weights using model.load_weights(). However, this approach requires that we reconstruct the model architecture separately using InceptionResNetV2(weights=None). This means that we first create an empty model with the same architecture as the original model, and then load only the weights from the saved model into this empty model.

By loading only the model weights and reconstructing the model architecture separately, we can avoid the computational overhead of reconstructing the entire model architecture when loading the saved model. This should result in faster loading times and better performance overall

Updated Code:

import tensorflow as tf
from tensorflow.keras.applications import InceptionResNetV2


def test_save_inception():
    model = InceptionResNetV2(weights=None)
    model_path = "models/inception"
    model.save(model_path)


def test_load_inception():
    model_path = "models/inception"
    model = tf.keras.models.load_model(model_path, compile=False)
    model.summary()


def test_load_inception_fast():
    model_path = "models/inception"
    weights_path = "models/inception_weights.h5"

    # create an empty model with the same architecture as the original model
    model = InceptionResNetV2(weights=None)

    # load only the model weights
    model.load_weights(weights_path)

    model.summary()

test_save_inception()
test_load_inception()
test_load_inception_fast()

**Comment by JustASquid** [comment]
Thanks for taking a look @NiharJani2002!

I should be more specific.

Because we are passing in the `InceptionResNetV2` as a custom object, it is my understanding that Keras should not need to reconstruct the entire model architecture from scratch. Instead, shouldn't it just create an instance using that constructor and load the weights, exactly as I am doing it in the "fast" version? And if not, why not?

**Comment by NiharJani2002** [comment]
@JustASquid.
You are correct when passing InceptionResNetV2 as a custom object to tf.keras.models.load_model(), Keras should create an instance of the InceptionResNetV2 model using the constructor and then load the weights, just as you are doing in the "fast" version of the code.

However, it seems that in practice, Keras is not actually doing this efficiently and is still reconstructing the entire model architecture from scratch when loading the saved model with a custom object. The reason for this behavior is likely due to the way Keras loads the custom object.

When a custom object is passed to tf.keras.models.load_model(), Keras first attempts to load the custom object using keras.utils.get_custom_objects(). If the custom object is not found in this dictionary, Keras will then attempt to import the custom object using keras.utils.get_registered_name(). Finally, if the custom object is still not found, Keras will raise a ValueError.

It's possible that this process of loading the custom object is what causes Keras to reconstruct the entire model architecture from scratch, even when a custom object is provided. This can be inefficient for large models like InceptionResNetV2, which can take a long time to reconstruct.

To avoid this overhead, it's best to load the model weights separately using model.load_weights() and then create an instance of the model with the same architecture using the constructor, as you are doing in the "fast" version of the code. This approach should be more efficient and result in faster loading times.

**Comment by JustASquid** [comment]
@NiharJani2002 

It would be nice to do the "fast approach" like you said. However, there is one major problem, and that is we do not know the constructor parameters to build the model correctly so that we can load the weights.

This is handled internally by keras with the `get_config()` and `from_config()` functions. It would be great to be able to do something like the following

```
model = InceptionResNetV2.from_config(config)
```
instead of 
```
model = InceptionResNetV2()
```

But of course, we would need to somehow load in the config ourselves.
Is there a way to extract only the `config` from the saved model?

**Comment by NiharJani2002** [comment]
@JustASquid 

Yes, you can extract the configuration of the model from the saved model file and use it to recreate the model with the same architecture as before, and then load the weights separately using model.load_weights().

To extract the configuration, you can use the get_config() method of the model instance that was saved. This method returns a dictionary containing the configuration information for the model. You can then use this dictionary to recreate the model using the from_config() method of the model class.

from keras.applications import InceptionResNetV2
import tensorflow as tf
import json

# Save the model
model = InceptionResNetV2(weights=None)
model.save("models/inception")

# Load the model
model_path = "models/inception"
with open(model_path + "/saved_model_config.json") as f:
    config = json.load(f)
model = InceptionResNetV2.from_config(config)
model.load_weights(model_path + "/variables/variables")

**Comment by JustASquid** [comment]
@NiharJani2002 
The saved_model_config.json file isn't saved by Keras as far as I can see.
saving the config as a separate JSON file manually would work but is error prone. The config should be saved somewhere in the saved model itself already anyway right? Ideally I'd like to pull that out.

**Comment by NiharJani2002** [comment]
@JustASquid 
You're correct that the saved_model_config.json file is not saved by Keras when using the model.save() method. Instead, Keras saves the model architecture in a saved_model.pb file, which is a protocol buffer file that contains a binary representation of the model graph.

To extract the configuration information from the saved_model.pb file, you can use the tf.compat.v1.graph_util.extract_sub_graph() function to extract the subgraph corresponding to the model's main computation graph, and then use the tf.compat.v1.graph_util.convert_variables_to_constants() function to convert the resulting subgraph to a TensorFlow GraphDef protocol buffer. You can then use the GraphDef to extract the model configuration information and recreate the model using the from_config() method.

Here's an example code snippet:

`
import tensorflow as tf
from keras.applications import InceptionResNetV2

# Save the model
model = InceptionResNetV2(weights=None)
model.save("models/inception")

# Load the model configuration
model_path = "models/inception"
with tf.compat.v1.Session() as sess:
    with tf.io.gfile.GFile(model_path + "/saved_model.pb", "rb") as f:
        graph_def = tf.compat.v1.GraphDef()
        graph_def.ParseFromString(f.read())
        subgraph_def = tf.compat.v1.graph_util.extract_sub_graph(graph_def, ["input_1", "predictions"])
        frozen_graph_def = tf.compat.v1.graph_util.convert_variables_to_constants(sess, subgraph_def, ["predictions"])
config = json.loads(frozen_graph_def.attr["model_config"].s.decode("utf-8"))
model = InceptionResNetV2.from_config(config)
model.load_weights(model_path + "/variables/variables")
`

In this example, we first save the model using model.save(), which saves the model architecture and weights to disk. We then load the model configuration by first reading in the saved_model.pb file using tf.io.gfile.GFile(). We then extract the subgraph corresponding to the model's main computation graph using tf.compat.v1.graph_util.extract_sub_graph(), and convert the resulting subgraph to a TensorFlow GraphDef protocol buffer using tf.compat.v1.graph_util.convert_variables_to_constants(). We then extract the model configuration information from the GraphDef using frozen_graph_def.attr["model_config"].s, which is a byte string that contains a JSON representation of the model configuration. Finally, we recreate the model using the from_config() method of the InceptionResNetV2 class, 
and load the weights separately using model.load_weights().

**Comment by JustASquid** [comment]
Hi @NiharJani2002 

Thanks for your suggestion.
Unfortunately it relies on deprecated functions, so I wouldn't be comfortable using it in production code. Is there any way to do this without deprecated functions? If not, I'd like to convert this to a feature request; Ideally Keras should be able to do this sort of thing automatically when the custom objects are passed to the `load_model()` function; A secondary solution would be a function like `keras.models.load_config()` which loads only the config for the model.

And finally, the code doesn't seem to work for me at all.
When I run this code on my M1, I get the following protobuf decode error during `graph_def.ParseFromString(f.read())`

```
../../../../../Library/Caches/pypoetry/virtualenvs/sorting-WRJTK5g2-py3.9/lib/python3.9/site-packages/google/protobuf/message.py:199: in ParseFromString
    return self.MergeFromString(serialized)
../../../../../Library/Caches/pypoetry/virtualenvs/sorting-WRJTK5g2-py3.9/lib/python3.9/site-packages/google/protobuf/internal/python_message.py:1128: in MergeFromString
    if self._InternalParse(serialized, 0, length) != length:
../../../../../Library/Caches/pypoetry/virtualenvs/sorting-WRJTK5g2-py3.9/lib/python3.9/site-packages/google/protobuf/internal/python_message.py:1195: in InternalParse
    pos = field_decoder(buffer, new_pos, end, self, field_dict)
../../../../../Library/Caches/pypoetry/virtualenvs/sorting-WRJTK5g2-py3.9/lib/python3.9/site-packages/google/protobuf/internal/decoder.py:726: in DecodeField
    if value._InternalParse(buffer, pos, new_pos) != new_pos:
../../../../../Library/Caches/pypoetry/virtualenvs/sorting-WRJTK5g2-py3.9/lib/python3.9/site-packages/google/protobuf/internal/python_message.py:1195: in InternalParse
    pos = field_decoder(buffer, new_pos, end, self, field_dict)
../../../../../Library/Caches/pypoetry/virtualenvs/sorting-WRJTK5g2-py3.9/lib/python3.9/site-packages/google/protobuf/internal/decoder.py:705: in DecodeRepeatedField
    if value.add()._InternalParse(buffer, pos, new_pos) != new_pos:
../../../../../Library/Caches/pypoetry/virtualenvs/sorting-WRJTK5g2-py3.9/lib/python3.9/site-packages/google/protobuf/internal/python_message.py:1195: in InternalParse
    pos = field_decoder(buffer, new_pos, end, self, field_dict)
../../../../../Library/Caches/pypoetry/virtualenvs/sorting-WRJTK5g2-py3.9/lib/python3.9/site-packages/google/protobuf/internal/decoder.py:860: in DecodeMap
    if submsg._InternalParse(buffer, pos, new_pos) != new_pos:
../../../../../Library/Caches/pypoetry/virtualenvs/sorting-WRJTK5g2-py3.9/lib/python3.9/site-packages/google/protobuf/internal/python_message.py:1181: in InternalParse
    (data, new_pos) = decoder._DecodeUnknownField(
../../../../../Library/Caches/pypoetry/virtualenvs/sorting-WRJTK5g2-py3.9/lib/python3.9/site-packages/google/protobuf/internal/decoder.py:961: in _DecodeUnknownField
    (data, pos) = _DecodeUnknownFieldSet(buffer, pos)
../../../../../Library/Caches/pypoetry/virtualenvs/sorting-WRJTK5g2-py3.9/lib/python3.9/site-packages/google/protobuf/internal/decoder.py:940: in _DecodeUnknownFieldSet
    (data, pos) = _DecodeUnknownField(buffer, pos, wire_type)
google.protobuf.message.DecodeError: Wrong wire type in tag.

**Comment by hertschuh** [comment]
Hi @JustASquid ,

`load_model` does something very different from just instantiating the model. There is definitely overhead when doing that.

However, Keras has a new saving format which reconstruct the model in a manner that is fairly similar from what you do in the constructor.

Here is how you can get the new format:

```py
def test_save_inception_v3():
    model_path = Path("models/inception.keras")
    model.save(model_path, save_format="keras_v3")
```

And then

```py
def test_load_inception_v3():
    model_path = Path("models/test_inception.keras")
    model = tf.keras.models.load_model(model_path, custom_objects={
        "InceptionResNetV2": InceptionResNetV2
    })
    model.summary()
```

I tested it, and on my machine the v3 format loads in 5.5 seconds as opposed to 26 seconds for the `load_model` version.
